### PR TITLE
[7.4 only] LPS-112473 Update Clay Select Tag to JSP+React

### PR DESCRIPTION
**Pull request by kresimir-coko**
Creating a new `SelectTag` implementing the Clay v3 version of the Select component.

This tag isn't being hydrated in this iteration, we've decided to add hydration when necessary.

## Steps to test:
- Add Clay Sample Widget to any page.
- Go to Form Elements tab
- Under the Select section

## Before:
image

## After:
image

### CODE REVIEW

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/SelectTag.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+	@Override
+	protected void writeDynamicAttributes() {
+	}
+
+	@Override
+	protected void writeIdAttribute() {
+	}
+
+	private static final String _ATTRIBUTE_NAMESPACE = "clay:selecttwo:";
```
Comment (jbalsas) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/SelectTag.java:182]: I guess this is just a test? 😂

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/SelectTag.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+		_disabled = false;
+		_label = null;
+		_multiple = false;
+		_name = null;
+		_options = null;
+	}
+
+	@Override
+	protected String getHydratedModuleName() {
+		return "frontend-taglib-clay/Select";
```
Comment (jbalsas) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/SelectTag.java:101]: I don't think we want to always hydrate a select. Only if we want to pass additional props, so maybe we can start with not hydrating at all and then add it in the near future.

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/SelectTag.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+	}
+
+	@Override
+	protected Map<String, Object> prepareProps(Map<String, Object> props) {
```
Comment (jbalsas) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/SelectTag.java:105]: Focus on the HTML generation now and forget about hydration to make this change simpler.

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/SelectTag.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+		JspWriter jspWriter = pageContext.getOut();
+
+		if (Validator.isNotNull(_label)) {
+			jspWriter.write("<label>");
+			jspWriter.write(_label);
+			jspWriter.write("</label>");
+		}
+
+		jspWriter.write("<select ");
```
Comment (jbalsas) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/SelectTag.java:138]: I guess we need to do this just because of the `<label>` so we can't use the `baseContainer`...

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/SelectTag.java (hunk 1 above)
Comment (kresimir-coko) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/SelectTag.java:182]: Yea, I had two select tags just to compare them easier 😄

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/SelectTag.java (hunk 2 above)
Comment (kresimir-coko) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/SelectTag.java:101]: Sure, sounds good 👍

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/SelectTag.java (hunk 4 above)
Comment (kresimir-coko) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/SelectTag.java:138]: Yea, looking at the Soy tag, it renders the following:
```html
<div class="form-group">
    <label>Label</label>

    <select class="form-control">
        <option />
        <option />
        <option />
        <option />
    </select>
</div>
```

If I made it select then we wouldn't be able to have the `<label>`, and probably not even the `<div class="form-group">`. I started with `baseContainer` being set to `select` but quickly saw that it won't work.

Comment (jbalsas) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/SelectTag.java:138]: Okay, okay, that's good, then.

**Comment by kresimir-coko** [comment]
I've pushed a commit that encapsulates the whole PR, I've tested it both with and without JS enabled, and it behaves as it's supposed to. I've added `cssClass` definition to the `tld` and deprecated attributes no longer used.

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/SelectTag.java [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+		jspWriter.write(" class=\"form-group\"");
+	}
+
+	@Override
+	protected void writeDynamicAttributes() {
+	}
+
+	@Override
+	protected void writeIdAttribute() {
+	}
```
Comment (julien) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/SelectTag.java:175]: @kresimir-coko just a doubt: is this intended? If we override these methods but leave them empty they won't do [this](https://github.com/liferay/liferay-portal/blob/2ba32f018d39b3eaacd32ff6c0fed7cfb806cff7/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/internal/servlet/taglib/BaseContainerTag.java#L387-L404)

Comment (kresimir-coko) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/SelectTag.java:175]: I think it's intended because we don't want an `id` on the container element (`<div>`), but we want it on the `<select>`

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/SelectTag.java:175]: ok thanks for clarifying, I didn't see [that](https://github.com/liferay-frontend/liferay-portal/pull/492/files/b5cde87d570a421d1474bc560f43b4766fb96e4f#diff-40c9b154810e3ef867cb61c2c11ad2accd0159d0e36350c91a6e7bc734dc7292R128),  now it makes sense

**Comment by kresimir-coko** [comment]
This is blocked by https://github.com/liferay/clay/issues/3796, so making this a Draft in the meanwhile.

**Review by markocikos — COMMENTED** [review]
The code as it is LGTM.

But, in the card tasks, we took the approach of letting go of Clay v2 markup and implementing based on Clay v3. As such, I think we should be consistent and make tag based on [SelectWithOption](https://github.com/liferay/clay/blob/master/packages/clay-form/src/SelectWithOption.tsx). More specifically, I mean we should not have this markup:
```
<div class="form-group">
    <label>Label</label>

    <select class="form-control">...</select>
</div>
```
and instead have only
```
<select>...</select>
```
With that approach, we would also deprecate the `label` attribute.

As far as I can see, `<clay:select>` is not used anywhere, except in sample-web, so the breaking compatibility aspect is minimal. I don't see a reason to keep the baggage of opinionated wrapper from v2. Also, this would simplify the server-side rendering, as attributes that are typically relayed to container will consistently do so.

@carloslancha @julien what do you think?

**Comment by julien** [comment]
@markocikos makes sense to me too

**Comment by carloslancha** [comment]
@markocikos that's exactly what I talked about with @kresimir-coko offline, I was going to write it down here but you were faster 👍 

We're not using `<clay:select>` anywhere but we don't know who is using it out there.

To me makes sense that `<clay:select>` only renders `<select><options></select>` but in that case:

- We are introducen a breaking change that we should document pretty well (we'd be getting rid of the label, it's not just about markup differences)
- We will be not being consistent between components, just as I commented [here](https://github.com/liferay/clay/issues/3796)

Could we move this conversation to that Clay issue so we could get all the context in the same place? :)

**Comment by kresimir-coko** [comment]
Marked as ready for review, as the component isn't used anywhere so we can adapt it later to whatever we decide on with the tag structure.

**Comment by jbalsas** [comment]
> We won't be consistent between components, just as I commented here

IMHO, we should be consistent within contexts, so:
- `<clay:select>` and `<clay:checkbox>` should behave similarly
- `<ClaySelect>` and `<ClayCheckbox>` should behave similarly
- `<clay:select>` and `<ClaySelect>` do not necessarily have to behave in the same way
- `<clay:checkbox>` and `<ClayCheckbox>` do not necessarily have to behave in the same way

There is an obvious problem with `<clay:select>` and `<ClaySelect>` not behaving identically, but there are also problems if they do, since their usage (and user) is radically different. 

We need to avoid hard set rules and evaluate in context, based on real usage and put cost-benefit in the balance.

> We're not using <clay:select> anywhere but we don't know who is using it out there.

Probably no-one. This is one where breaking it is specially cheap if it helps us. We just need to make sure to document it even as an excess of caution.

So, in summary, `<clay:select>` should behave the way it makes sense for a JSP tags, and stay consistent to how we develop in JSPs. The conversation about `<ClaySelect>` can continue at https://github.com/liferay/clay/issues/3796 independently.

**Comment by markocikos** [comment]
So @kresimir-coko, I guess your task is to find out if we are using jsp `<select>`s in all sorts of different contexts, or in high majority with  v2 markup (div with form-group class, label etc.)

**Comment by jbalsas** [comment]
> So @kresimir-coko, I guess your task is to find out if we are using jsp <select>s in all sorts of different contexts, or in high majority with v2 markup (div with form-group class, label etc.)

The main use case would be to replace `aui:select`, so that's the main pattern we have to worry about.

**Comment by kresimir-coko** [comment]
Investigative work, I see, not terrible 👁️

**Comment by kresimir-coko** [comment]
# Investigation Report 

## Wrapper variants
- In 100% of the usages the tag has `<aui:option>` tag in some form as a child of it.
- In most use cases the `label` attribute is defined with a value, or with an empty string in some cases.

Here's a list of a few variations of usages of `<aui:select>` tag:

- In some cases the `options` are wrapped in a `<optgroup>` HTML tag
- In some cases the `<aui:select>` has a custom wrapper that handles hiding
- In some cases the `<aui:select>` is a child of `<aui:form>`
- Otherwise, the `<aui:select>` is a child of some other tag, like `<liferay-frontend-fieldset>`, or something arbitrary like `<li>`, a div with an arbitrary class, etc.
- In 4 cases, `<aui:select>` is a child of `<div class="form-group-item">`
- In 4 cases, `<aui:select>` is a child of `<div class="form-group">`

## Attributes used that the `<clay:select>` server-side doesn't have
- `bean`
- `changesContext`
- `helpMessage`
- `inlineField`
- `inlineLabel`
- `listType`
- `listTypeFieldName`
- `model`
- `onChange`
- `readOnly`
- `showEmptyOption`

**Comment by markocikos** [comment]
@kresimir-coko This is raw clue data. The job of the investigator is to deduct a conclusion :monocle_face:  What markup makes more sense to use?

**Comment by kresimir-coko** [comment]
Crunching clues... Deducting information... Drinking coffee... 

The markup that makes the most sense is to **just have a `<select>` element** in the `<clay:select>` tag, because 99% of current examples could get messed up by adding a `<div class="form-group">` around already incorporated `<aui:select>` tags

**Comment by kresimir-coko** [comment]
Upon further investigation, I looked at the DOM structure of `<aui:select>`, specifically this one found in `announcements-web`, thing to note here is that the `<aui:select>` doesn't have any direct parent elements that would render any elements directly around it: 
```jsp
<aui:select name="priority">
	<aui:option label="normal" selected="<%= (entry != null) && (entry.getPriority() == 0) %>" value="0" />
	<aui:option label="important" selected="<%= (entry != null) && (entry.getPriority() == 1) %>" value="1" />
</aui:select>
```

It renders the following:
```html
<div class="form-group input-select-wrapper">
	<label class="control-label" for="_com_liferay_announcements_web_portlet_AnnouncementsPortlet_priority">Priority</label>

	<select class="form-control" id="_com_liferay_announcements_web_portlet_AnnouncementsPortlet_priority" name="_com_liferay_announcements_web_portlet_AnnouncementsPortlet_priority">
		<option class="" selected="" value="0">Normal</option>
		<option class="" value="1">Important</option>
	</select>
</div>
```

Looking at this, and looking at the implementation of `<aui:select>` we can see that the tag **renders a `<div class="form-group input-select-wrapper">` by default, and a `<label class="control-label">` always**, even if `label` attribute isn't defined, it will use `name` to determine the value of the `label`.

So, if we want to make replacing old usages easier, we should maintain the same DOM structure, but if we want to have a more flexible component/tag we should make it render only a `<select>` element.

What are your thoughts on this @jbalsas @carloslancha @julien @markocikos @wincent

**Comment by carloslancha** [comment]
@kresimir-coko I think the question is: Is there any case where we're not rendering a form-group + label?

**Comment by kresimir-coko** [comment]
To not waste months trying to manually find every usage of `<aui:select>`, it kinda makes sense that only the `wrapperCssClass` prop has influence on the wrapper element, and it probably just adds classes to the `form-group`.

As far as a case where a label isn't rendered, because a `name` is required, and there's this block of code that makes sure the `label` prop is always defined:
```java
if (label == null) {
	label = TextFormatter.format(name, TextFormatter.K);
}
```

As I can't find the start and end JSPs, they're like 9 years old idk, I can't confirm this, but it makes sense to always render it.

**Comment by kresimir-coko** [comment]
@jbalsas I'm lost here, not sure how to proceed, above I've done the investigation and deductive work, but we still haven't reached a decision, so I'm kinda stuck.

**Comment by jbalsas** [comment]
Please, sync with @markocikos and @carloslancha to figure out how to proceed.

**Comment by markocikos** [comment]
Well, if there are good reasons to have both ways, why not both?

I would suggest that we have two tags:
1. `<clay:select>` that renders only `<select>`
1. `<clay:form-select>` that renders select in the `form-group` wrapper, with label etc.

What do you think?

**Comment by kresimir-coko** [comment]
I think that's not a bad idea, but would require maintaining 2 tags, also, there's a multiselect tag that needs to be done after this, so that one will probably have those 2 variants as well, meaning 4 tags instead of 2.

Is it possible to have `<clay:form-select>` call `<clay:select>`, that would eliminate the concern of maintaining them separately?

**Comment by carloslancha** [comment]
@kresimir-coko based on your research I'd say:

- Add a `containerCssClass` param to the tag
- SelectTag should render a label and a select with options, both wrapped inside a div with `form-group` and the passed `containerCssClass`.

And that's it, we can add more stuff if we need it once we start replacing.

> I would suggest that we have two tags:
<clay:select> that renders only <select>
<clay:form-select> that renders select in the form-group wrapper, with label etc.

I'd prefer not to add more tags if we're not 100% sure they're gonna be needed, and based on Kreso's research, what we need here is a wrapped select in all the cases, so I' just go with `<clay:select>` for it.

**Comment by kresimir-coko** [comment]
Okay, I'll go with your suggestion @carloslancha and then we can work towards extending/trimming it once we encounter some problems replacing `<aui:select>` which will definitely creep up.

**Comment by markocikos** [comment]
How about an attribute that adds wrapper:
```
<clay:select
    formWrapper = 'true'
>
```
?

**Comment by carloslancha** [comment]
> How about an attribute that adds wrapper:

I'd add it only when we face a use case where we really need it

**Comment by kresimir-coko** [comment]
@markocikos 
> How about an attribute that adds wrapper

Wouldn't that introduce a lot of other conditionals? Because a label won't be able to be present without a wrapper, and there's like a 1% chance anyone will want a Select without a Label.

It's a good idea, but probably better to save it for later, like @carloslancha said

**Comment by markocikos** [comment]
Fair enough, we can always iterate later on :+1:

**Comment by jbalsas** [comment]
> [...] there's a multiselect tag [...]

Just FYI, `multi-select` has nothing to do with `select`. At least not in the sense that seemed to imply. You could've used any other component like `checkbox`, `input`, `radio`, `datepicker`, `timepicker`...

**Comment by kresimir-coko** [comment]
@carloslancha I've pushed a new commit that adds the `containerCssClass` prop, and uses it to showcase that it works. Let me know if that's how you imagined it would be implemented.

**Comment by carloslancha** [comment]
Yep @kresimir-coko exactly.

This pr is not horrible 😉

**Comment by carloslancha** [comment]
ci:test:sf

**Comment by carloslancha** [comment]
ci:test:relevant

**Comment by pat270** [comment]
Expanding on what @carloslancha said in https://github.com/liferay-frontend/liferay-portal/pull/492#issuecomment-728894175. Whatever container param name we decide to pick should be consistent across all our inputs. The current (old aui:input) tag uses `wrapperCssClass`.

We also should give  the option of not outputting a `label`. No label select elements are commonly used in navigations. One example in portal is in the Product Menu Page Tree section.

Some documentation for `aui:input` https://github.com/liferay/liferay-frontend-guidelines/blob/master/dxp/taglib_aui_input.md#text.

**Comment by carloslancha** [comment]
Yep, thx @pat270 !

>The current (old aui:input) tag uses wrapperCssClass.

I said `contentCssClass` as an example, I've no preference here so we can name it `wrapperCssClass` to follow the same naming or use whatever name makes more sense.

>We also should give the option of not outputting a label. No label select elements are commonly used in navigations. One example in portal is in the Product Menu Page Tree section.

This pr support it already, if no `label` is passed no `label` is rendered 😄

**Comment by carloslancha** [comment]
After a second look I think `containerCssClass` is a good name since we do have some `containerElement` attributes in clay taglib. I'm forwarding this, if anyone has other opinion please feel free to comment here and we can address it in further prs.

Thx!

**Comment by carloslancha** [comment]
ci:forward
